### DISCUSSION TITLE
4.0.3 Event Rules - no scripts in dropdown

### DISCUSSION THREAD

**Discussion by ziggekatten** [netbox-community/netbox#16361] [Help Wanted!] [answered]
Upgraded my dev environment to 4.0.0 from 3.7.5, and everything went smooth. All my custom scripts are working as they should, and so does the existing migrated event rules from the 3.7.5 version. 
However, trying to add an new event rule of type script, there is no scripts in the drop down:
![image](https://github.com/netbox-community/netbox/assets/55629049/22b31076-259d-4654-bbb0-edce4ee3e908)

Went along and upgraded to 4.0,3 to ensure that this is not an known bug, but no difference. journalctl shows below error when I click on the dropdown:

```
May 31 08:52:29 SELI148185 gunicorn[754762]: Internal Server Error: /api/extras/scripts/
May 31 08:52:29 SELI148185 gunicorn[754762]: Traceback (most recent call last):
May 31 08:52:29 SELI148185 gunicorn[754762]:   File "/opt/netbox-4.0.3/venv/lib/python3.10/site-packages/django/core/handlers/exception.py", line 55, in inner
May 31 08:52:29 SELI148185 gunicorn[754762]:     response = get_response(request)
May 31 08:52:29 SELI148185 gunicorn[754762]:   File "/opt/netbox-4.0.3/venv/lib/python3.10/site-packages/django/core/handlers/base.py", line 220, in _get_response
May 31 08:52:29 SELI148185 gunicorn[754762]:     response = response.render()
May 31 08:52:29 SELI148185 gunicorn[754762]:   File "/opt/netbox-4.0.3/venv/lib/python3.10/site-packages/django/template/response.py", line 114, in render
May 31 08:52:29 SELI148185 gunicorn[754762]:     self.content = self.rendered_content
May 31 08:52:29 SELI148185 gunicorn[754762]:   File "/opt/netbox-4.0.3/venv/lib/python3.10/site-packages/rest_framework/response.py", line 74, in rendered_content
May 31 08:52:29 SELI148185 gunicorn[754762]:     ret = renderer.render(self.data, accepted_media_type, context)
May 31 08:52:29 SELI148185 gunicorn[754762]:   File "/opt/netbox-4.0.3/venv/lib/python3.10/site-packages/rest_framework/renderers.py", line 99, in render
May 31 08:52:29 SELI148185 gunicorn[754762]:     ret = json.dumps(
May 31 08:52:29 SELI148185 gunicorn[754762]:   File "/opt/netbox-4.0.3/venv/lib/python3.10/site-packages/rest_framework/utils/json.py", line 25, in dumps
May 31 08:52:29 SELI148185 gunicorn[754762]:     return json.dumps(*args, **kwargs)
May 31 08:52:29 SELI148185 gunicorn[754762]:   File "/usr/lib/python3.10/json/__init__.py", line 238, in dumps
May 31 08:52:29 SELI148185 gunicorn[754762]:     **kw).encode(obj)
May 31 08:52:29 SELI148185 gunicorn[754762]:   File "/usr/lib/python3.10/json/encoder.py", line 199, in encode
May 31 08:52:29 SELI148185 gunicorn[754762]:     chunks = self.iterencode(o, _one_shot=True)
May 31 08:52:29 SELI148185 gunicorn[754762]:   File "/usr/lib/python3.10/json/encoder.py", line 257, in iterencode
May 31 08:52:29 SELI148185 gunicorn[754762]:     return _iterencode(o, 0)
May 31 08:52:29 SELI148185 gunicorn[754762]:   File "/opt/netbox-4.0.3/venv/lib/python3.10/site-packages/rest_framework/utils/encoders.py", line 67, in default
May 31 08:52:29 SELI148185 gunicorn[754762]:     return super().default(obj)
May 31 08:52:29 SELI148185 gunicorn[754762]:   File "/usr/lib/python3.10/json/encoder.py", line 179, in default
May 31 08:52:29 SELI148185 gunicorn[754762]:     raise TypeError(f'Object of type {o.__class__.__name__} '
May 31 08:52:29 SELI148185 gunicorn[754762]: TypeError: Object of type StringVar is not JSON serializable
```

Am I missing something here?

**Comment by candlerb** [comment]
What happens if you go to Customizations > Scripts? Do you see the scripts listed, but with no "Run script" button?

That's what happened to me after migration from 3.7. Somehow the scripts were not registered in the database. The solution was to delete and then re-upload every script (which was tedious)

But now if I try to add an event rule, I *do* see them all listed:

![image](https://github.com/netbox-community/netbox/assets/44789/cc4030de-3c4b-4b9c-b1b6-b46061c013ea)

**Reply by ziggekatten** [reply]
Yes, I see all scripts and I can run them. I'm currently trying to delete my scripts and re-add them, but when I try to delete:

![image](https://github.com/netbox-community/netbox/assets/55629049/97b464f9-6f2b-4f60-b020-ecdc1b1e74b7)

**Comment by ziggekatten** [comment, marked as answer]
Using nbshell I could delete all my scripts and event rules, and now I can create things again. Something borked something during upgrade......

**Comment by llamafilm** [comment]
Did you happen to save the output from `upgrade.sh` when upgrading from v3?  It should show something about the script migration.  I had a similar issue, reported in #16218, that might be related.  If the script file is not present when that migration is run (for example if they have not yet been synced from a remote data source), the migration won't work right.

**Comment by ziggekatten** [comment]
Yes, i suspect that missing files where the culprit as I use links to my dev area instead of copying files
